_cheetah.c

```diff
@@ -0,0 +1,106 @@
+#include <Python.h>
+
+#if PY_MAJOR_VERSION >= 3
+#define IF_PY3(three, two) (three)
+#else
+#define IF_PY3(three, two) (two)
+#endif
+
+static PyObject* NotFound;
+static PyObject* _builtins_module;
+
+
+static PyObject* _vfsl(char* key, PyObject* selfobj, PyObject* ns) {
+    if (PyObject_HasAttrString(selfobj, key)) {
+        return PyObject_GetAttrString(selfobj, key);
+    } else if (PyMapping_HasKeyString(ns, key)) {
+        return PyMapping_GetItemString(ns, key);
+    } else {
+        PyObject* fmt = PyUnicode_FromString("Cannot find '{0}'");
+        PyObject* fmted = PyObject_CallMethod(fmt, "format", IF_PY3("y", "s"), key);
+        PyErr_SetObject(NotFound, fmted);
+        Py_XDECREF(fmted);
+        Py_XDECREF(fmt);
+        return NULL;
+    }
+}
+
+static PyObject* value_from_search_list(PyObject* _, PyObject* args) {
+    char* key;
+    PyObject* selfobj;
+    PyObject* ns;
+
+    if (!PyArg_ParseTuple(args, "sOO", &key, &selfobj, &ns)) {
+        return NULL;
+    }
+
+    return _vfsl(key, selfobj, ns);
+}
+
+static PyObject* value_from_frame_or_search_list(PyObject* _, PyObject* args) {
+    char* key;
+    PyObject* locals;
+    PyObject* globals;
+    PyObject* selfobj;
+    PyObject* ns;
+
+    if (!PyArg_ParseTuple(args, "sOOOO", &key, &locals, &globals, &selfobj, &ns)) {
+        return NULL;
+    }
+
+    if (PyMapping_HasKeyString(locals, key)) {
+        return PyMapping_GetItemString(locals, key);
+    } else if (PyMapping_HasKeyString(globals, key)) {
+        return PyMapping_GetItemString(globals, key);
+    } else if (PyObject_HasAttrString(_builtins_module, key)) {
+        return PyObject_GetAttrString(_builtins_module, key);
+    } else {
+        return _vfsl(key, selfobj, ns);
+    }
```

Comment:
I guess I could gobble the exception like the implementation of HasAttrString does